crates/wasmtime/src/signatures.rs

```diff
@@ -0,0 +1,185 @@
+//! Implement a registry of function signatures, for fast indirect call
+//! signature checking.
+
+use std::collections::{hash_map::Entry, HashMap};
+use std::convert::TryFrom;
+use wasmtime_environ::entity::PrimaryMap;
+use wasmtime_environ::wasm::{SignatureIndex, WasmFuncType};
+use wasmtime_runtime::{VMSharedSignatureIndex, VMTrampoline};
+
+/// Represents a mapping of shared signature index to trampolines.
+///
+/// This is used in various places to store trampolines associated with shared
+/// signature indexes.
+///
+/// As multiple trampolines may exist for a single signature, the map entries
+/// are internally reference counted.
+#[derive(Default)]
+pub struct TrampolineMap(HashMap<VMSharedSignatureIndex, (usize, VMTrampoline)>);
+
+impl TrampolineMap {
+    /// Inserts a trampoline into the map.
+    pub fn insert(&mut self, index: VMSharedSignatureIndex, trampoline: VMTrampoline) {
+        let entry = match self.0.entry(index) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => e.insert((0, trampoline)),
+        };
+
+        // Increment the ref count
+        entry.0 += 1;
+    }
+
+    /// Gets a trampoline from the map.
+    pub fn get(&self, index: VMSharedSignatureIndex) -> Option<VMTrampoline> {
+        self.0.get(&index).map(|(_, trampoline)| *trampoline)
+    }
+
+    /// Iterates the shared signature indexes stored in the map.
+    ///
+    /// A shared signature index will be returned by the iterator for every
+    /// trampoline registered for that index, so duplicates may be present.
+    ///
+    /// This iterator can be used for deregistering signatures with the
+    /// signature registry.
+    pub fn indexes<'a>(&'a self) -> impl Iterator<Item = VMSharedSignatureIndex> + 'a {
+        self.0
+            .iter()
+            .flat_map(|(index, (count, _))| std::iter::repeat(*index).take(*count))
+    }
```

Comment:
If this function is only used for deregistration, could this perhaps return an iterator over the signature index and the reference count? That way the deregistration wouldn't have to do lookup multiple times in a row